--- a/src/Makefile
+++ b/src/Makefile
@@ -2,7 +2,6 @@
 # Copyright (C) 2025 StarField Xu <air_jinkela@163.com>
 
 ccflags-y += -Wall
-ccflags-y += -Iinclude/linux/
 
 ifeq ($(CONFIG_RTL8372N_DSA_DEBUG),y)
 ccflags-y += -DDEBUG
@@ -19,6 +18,7 @@ endif
 rtl8372n_dsa-objs += rtl837x_dbg.o
 rtl8372n_dsa-objs += rtl837x_mdio.o
 rtl8372n_dsa-objs += rtl837x_common.o
+rtl8372n_dsa-objs += rtl837x_common_l2.o
 
 ifeq ($(CONFIG_RTL837X_PHY_PATCH),y)
 ccflags-y += -DRTL837X_PHY_PATCH
--- a/src/rtl837x_common.c
+++ b/src/rtl837x_common.c
@@ -685,8 +685,6 @@ int rtl837x_vlan_get(struct rtl837x_priv
 	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
 		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
 		  0, 1000);
-
-	ret = rtl837x_reg_write(priv, RTL8373_ITA_WRITE_DATA0_ADDR(0), vlan->val);
 	if (ret)
 		return ret;
 
--- /dev/null
+++ b/src/rtl837x_common_l2.c
@@ -0,0 +1,455 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright (C) 2025 StarField Xu <air_jinkela@163.com>
+ */
+#include <linux/version.h>
+
+#include "rtl837x.h"
+
+#define RTL837X_LUT_TABLE_SIZE         (3)
+
+/*  unicast field define  */
+#define RTL837X_L2_UC_D0_MAC5_MSK		GENMASK(7, 0)
+#define RTL837X_L2_UC_D0_MAC4_MSK		GENMASK(15, 8)
+#define RTL837X_L2_UC_D0_MAC3_MSK		GENMASK(23, 16)
+#define RTL837X_L2_UC_D0_MAC2_MSK		GENMASK(31, 24)
+#define RTL837X_L2_UC_D1_MAC1_MSK		GENMASK(7, 0)
+#define RTL837X_L2_UC_D1_MAC0_MSK		GENMASK(15, 8)
+
+#define RTL837X_L2_UC_D1_CVIDFID_MSK	GENMASK(27, 16)
+#define RTL837X_L2_UC_D1_L3LOOKUP_MSK	GENMASK(28, 28)
+#define RTL837X_L2_UC_D1_IVL_MSK		GENMASK(29, 29)
+
+#define RTL837X_L2_UC_D1_SPORT_LO_MSK	GENMASK(31, 30)
+#define  RTL837X_L2_UC_D2_SPORT_HI_MSK	GENMASK(1, 0)
+#define  RTL837X_L2_UC_SPORT_GET(_D1, _D2) \
+			((FIELD_GET(RTL837X_L2_UC_D1_SPORT_LO_MSK, _D1))| \
+			 (FIELD_GET(RTL837X_L2_UC_D2_SPORT_HI_MSK, _D2) << 2))
+
+#define RTL837X_L2_UC_D2_AGE_MSK		GENMASK(4, 2)
+#define RTL837X_L2_UC_D2_AUTH_MSK		GENMASK(5, 5)
+#define RTL837X_L2_UC_D2_STATIC_MSK		GENMASK(16, 16)
+
+/*  multicast field define  */
+#define RTL837X_L2_MC_D0_MAC5_MSK		GENMASK(7, 0)
+#define RTL837X_L2_MC_D0_MAC4_MSK		GENMASK(15, 8)
+#define RTL837X_L2_MC_D0_MAC3_MSK		GENMASK(23, 16)
+#define RTL837X_L2_MC_D0_MAC2_MSK		GENMASK(31, 24)
+#define RTL837X_L2_MC_D1_MAC1_MSK		GENMASK(7, 0)
+#define RTL837X_L2_MC_D1_MAC0_MSK		GENMASK(15, 8)
+
+#define RTL837X_L2_MC_D1_CVIDFID_MSK	GENMASK(27, 16)
+#define RTL837X_L2_MC_D1_L3LOOKUP_MSK	GENMASK(28, 28)
+#define RTL837X_L2_MC_D1_IVL_MSK		GENMASK(29, 29)
+
+#define RTL837X_L2_MC_D1_MBR_LO_MSK	    GENMASK(31, 30)
+#define  RTL837X_L2_MC_D2_MBR_HI_MSK	GENMASK(7, 0)
+#define  RTL837X_L2_MC_MBR_GET(_D1, _D2) \
+            ((FIELD_GET(RTL837X_L2_MC_D1_MBR_LO_MSK, _D1))| \
+             (FIELD_GET(RTL837X_L2_MC_D2_MBR_HI_MSK, _D2) << 2))
+
+#define RTL837X_L2_MC_D2_IGMPIDX_MSK	GENMASK(15, 8)
+#define RTL837X_L2_MC_D2_IGMPASIC_MSK	GENMASK(16, 16)
+
+/*  L3 field define  */
+/* this chip support l3? Interesting */
+#define RTL837X_L3_D0_SIP_MSK           GENMASK(31, 0)
+#define RTL837X_L3_D1_DIP_MSK           GENMASK(27, 0)
+#define RTL837X_L3_D1_L3LOOKUP_MSK      GENMASK(28, 28)
+
+#define RTL837X_L3_D1_MBR_LO_MSK	    GENMASK(31, 30)
+#define  RTL837X_L3_D2_MBR_HI_MSK	GENMASK(7, 0)
+#define  RTL837X_L3_MBR_GET(_D1, _D2) \
+            ((FIELD_GET(RTL837X_L3_D1_MBR_LO_MSK, _D1))| \
+             (FIELD_GET(RTL837X_L3_D2_MBR_HI_MSK, _D2) << 2))
+
+#define RTL837X_L3_D2_IGMPIDX_MSK	GENMASK(15, 8)
+#define RTL837X_L3_D2_IGMPASIC_MSK	GENMASK(16, 16)
+
+static void rtl837x_l2_data_to_uc(const u32 *data, struct rtl837x_l2_uc *uc)
+{
+	uc->key.mac_addr[5] = FIELD_GET(RTL837X_L2_UC_D0_MAC5_MSK, data[0]);
+	uc->key.mac_addr[4] = FIELD_GET(RTL837X_L2_UC_D0_MAC4_MSK, data[0]);
+	uc->key.mac_addr[3] = FIELD_GET(RTL837X_L2_UC_D0_MAC3_MSK, data[0]);
+	uc->key.mac_addr[2] = FIELD_GET(RTL837X_L2_UC_D0_MAC2_MSK, data[0]);
+	uc->key.mac_addr[1] = FIELD_GET(RTL837X_L2_UC_D1_MAC1_MSK, data[1]);
+	uc->key.mac_addr[0] = FIELD_GET(RTL837X_L2_UC_D1_MAC0_MSK, data[1]);
+
+	uc->key.vid_fid = FIELD_GET(RTL837X_L2_UC_D1_CVIDFID_MSK, data[1]);
+	// shoule never be set in uc entry
+	// uc->l3lookup = FIELD_GET(RTL837X_L2_UC_D1_L3LOOKUP_MSK, data[1]);
+	uc->key.ivl = FIELD_GET(RTL837X_L2_UC_D1_IVL_MSK, data[1]);
+
+	uc->port = RTL837X_L2_UC_SPORT_GET(data[1], data[2]);
+
+	uc->age = FIELD_GET(RTL837X_L2_UC_D2_AGE_MSK, data[2]);
+	uc->auth = FIELD_GET(RTL837X_L2_UC_D2_AUTH_MSK, data[2]);
+	uc->is_static = FIELD_GET(RTL837X_L2_UC_D2_STATIC_MSK, data[2]);
+}
+
+static void rtl837x_l2_uc_to_data(const struct rtl837x_l2_uc *uc, u32 *data)
+{
+	memset(data, 0, RTL837X_LUT_TABLE_SIZE * sizeof(*data));
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_UC_D0_MAC5_MSK, uc->key.mac_addr[5]);
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_UC_D0_MAC4_MSK, uc->key.mac_addr[4]);
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_UC_D0_MAC3_MSK, uc->key.mac_addr[3]);
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_UC_D0_MAC2_MSK, uc->key.mac_addr[2]);
+	data[1] |=
+		FIELD_PREP(RTL837X_L2_UC_D1_MAC1_MSK, uc->key.mac_addr[1]);
+	data[1] |=
+		FIELD_PREP(RTL837X_L2_UC_D1_MAC0_MSK, uc->key.mac_addr[0]);
+
+	data[1] |= 
+		FIELD_PREP(RTL837X_L2_UC_D1_CVIDFID_MSK, uc->key.vid_fid);
+	data[1] |= 
+		FIELD_PREP(RTL837X_L2_UC_D1_IVL_MSK, uc->key.ivl);
+	// shoule never be set in uc entry
+	data[1] |= 
+		FIELD_PREP(RTL837X_L2_UC_D1_L3LOOKUP_MSK, 0);
+
+	data[1] |=
+		FIELD_PREP(RTL837X_L2_UC_D1_SPORT_LO_MSK, uc->port & 0x3);
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_UC_D2_SPORT_HI_MSK, (uc->port >> 2) & 0x3);
+
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_UC_D2_STATIC_MSK, uc->is_static);
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_UC_D2_AGE_MSK, uc->age);
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_UC_D2_AUTH_MSK, uc->auth);
+}
+
+
+static void rtl837x_l2_data_to_mc(const u32 *data, struct rtl837x_l2_mc *mc)
+{
+	mc->key.mac_addr[5] = FIELD_GET(RTL837X_L2_MC_D0_MAC5_MSK, data[0]);
+	mc->key.mac_addr[4] = FIELD_GET(RTL837X_L2_MC_D0_MAC4_MSK, data[0]);
+	mc->key.mac_addr[3] = FIELD_GET(RTL837X_L2_MC_D0_MAC3_MSK, data[0]);
+	mc->key.mac_addr[2] = FIELD_GET(RTL837X_L2_MC_D0_MAC2_MSK, data[0]);
+	mc->key.mac_addr[1] = FIELD_GET(RTL837X_L2_MC_D1_MAC1_MSK, data[1]);
+	mc->key.mac_addr[0] = FIELD_GET(RTL837X_L2_MC_D1_MAC0_MSK, data[1]);
+
+	mc->key.vid_fid = FIELD_GET(RTL837X_L2_MC_D1_CVIDFID_MSK, data[1]);
+	// shoule never be set in mc entry
+	// mc->l3lookup    = FIELD_GET(RTL837X_L2_MC_D1_L3LOOKUP_MSK, data[1]);
+	mc->key.ivl     = FIELD_GET(RTL837X_L2_MC_D1_IVL_MSK, data[1]);
+
+	mc->mbr = RTL837X_L2_MC_MBR_GET(data[1], data[2]);
+
+	mc->igmp_asic = FIELD_GET(RTL837X_L2_MC_D2_IGMPASIC_MSK, data[2]);
+	mc->igmp_idx  = FIELD_GET(RTL837X_L2_MC_D2_IGMPIDX_MSK, data[2]);
+}
+
+static void rtl837x_l2_mc_to_data(const struct rtl837x_l2_mc *mc, u32 *data)
+{
+	memset(data, 0, RTL837X_LUT_TABLE_SIZE * sizeof(*data));
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_MC_D0_MAC5_MSK, mc->key.mac_addr[5]);
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_MC_D0_MAC4_MSK, mc->key.mac_addr[4]);
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_MC_D0_MAC3_MSK, mc->key.mac_addr[3]);
+	data[0] |=
+		FIELD_PREP(RTL837X_L2_MC_D0_MAC2_MSK, mc->key.mac_addr[2]);
+	data[1] |=
+		FIELD_PREP(RTL837X_L2_MC_D1_MAC1_MSK, mc->key.mac_addr[1]);
+	data[1] |=
+		FIELD_PREP(RTL837X_L2_MC_D1_MAC0_MSK, mc->key.mac_addr[0]);
+
+	data[1] |= 
+		FIELD_PREP(RTL837X_L2_MC_D1_CVIDFID_MSK, mc->key.vid_fid);
+	data[1] |= 
+		FIELD_PREP(RTL837X_L2_MC_D1_IVL_MSK, mc->key.ivl);
+
+	// shoule never be set in mc entry
+	data[1] |= 
+		FIELD_PREP(RTL837X_L2_MC_D1_L3LOOKUP_MSK, 0);
+
+	data[1] |=
+		FIELD_PREP(RTL837X_L2_MC_D1_MBR_LO_MSK, mc->mbr & 0x3);
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_MC_D2_MBR_HI_MSK, (mc->mbr >> 2) & 0xFF);
+
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_MC_D2_IGMPIDX_MSK, mc->igmp_idx);
+	data[2] |=
+		FIELD_PREP(RTL837X_L2_MC_D2_IGMPASIC_MSK, mc->igmp_asic);
+}
+
+static void rtl837x_data_to_l3(const u32 *data, struct rtl837x_l3 *l3)
+{
+	l3->sip = FIELD_GET(RTL837X_L3_D0_SIP_MSK, data[0]);
+	l3->dip = FIELD_GET(RTL837X_L3_D1_DIP_MSK, data[1]);
+
+	l3->l3lookup = FIELD_GET(RTL837X_L3_D1_L3LOOKUP_MSK, data[1]);
+
+	l3->mbr = RTL837X_L3_MBR_GET(data[1], data[2]);
+
+	l3->igmp_idx = FIELD_GET(RTL837X_L3_D2_IGMPIDX_MSK, data[2]);
+	l3->igmp_asic = FIELD_GET(RTL837X_L3_D2_IGMPASIC_MSK, data[2]);
+}
+
+static void rtl837x_l3_to_data(const struct rtl837x_l3 *l3, u32 *data)
+{
+	memset(data, 0, RTL837X_LUT_TABLE_SIZE * sizeof(*data));
+	data[0] |=
+		FIELD_PREP(RTL837X_L3_D0_SIP_MSK, l3->sip);
+	data[1] |=
+		FIELD_PREP(RTL837X_L3_D1_DIP_MSK, l3->dip);
+	data[1] |=
+		FIELD_PREP(RTL837X_L3_D1_L3LOOKUP_MSK, l3->l3lookup);
+
+	data[1] |=
+		FIELD_PREP(RTL837X_L3_D1_MBR_LO_MSK, l3->mbr & 0x3);
+	data[2] |=
+		FIELD_PREP(RTL837X_L3_D2_MBR_HI_MSK, (l3->mbr >> 2) & 0xFF);
+
+	data[2] |=
+		FIELD_PREP(RTL837X_L3_D2_IGMPIDX_MSK, l3->igmp_idx);
+	data[2] |=
+		FIELD_PREP(RTL837X_L3_D2_IGMPASIC_MSK, l3->igmp_asic);
+}
+
+// TODO: we need a lock to make sure the ITA read/write is atomic
+
+int rtl837x_lut_query(struct rtl837x_priv *priv, 
+                          enum rtl837x_l2_method method,
+                          struct rtl837x_lut_entry *entry)
+{
+	int ret;
+	u32 tmp, cmd;
+	u32 tb_data[RTL837X_LUT_TABLE_SIZE] = {0};
+
+	// check busy
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
+		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
+		  10, 10000);
+	if (ret)
+		return ret;
+
+	ret = rtl837x_reg_bits_write(priv, RTL8373_ITA_L2_CTRL_ADDR,
+					  RTL8373_ITA_L2_CTRL_READ_MTHD_MASK, method);
+	if (ret)
+		return ret;
+
+	switch (method)
+	{
+	case LUT_READ_METHOD_ADDRESS:
+	case LUT_READ_METHOD_NEXT_ADDRESS:
+	case LUT_READ_METHOD_NEXT_L2UC:
+	case LUT_READ_METHOD_NEXT_L2MC:
+	case LUT_READ_METHOD_NEXT_L3MC:
+	case LUT_READ_METHOD_NEXT_L2L3MC:
+		cmd = FIELD_PREP(RTL8373_ITA_CTRL0_TBL_ADDR_MASK, entry->addr) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_TYPE_MASK, TB_TARGET_L2) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_ACT_MASK, TB_OP_READ) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK, TB_EXECUTE);
+		break;
+	case LUT_READ_METHOD_MAC:
+		rtl837x_l2_uc_to_data(&entry->uc, tb_data);
+		for(int i=0; i<RTL837X_LUT_TABLE_SIZE; i++)
+		{
+			ret = rtl837x_reg_write(priv, RTL8373_ITA_WRITE_DATA0_ADDR(i), tb_data[i]);
+			if (ret)
+				return ret;
+		}
+		cmd = FIELD_PREP(RTL8373_ITA_CTRL0_TLB_TYPE_MASK, TB_TARGET_L2) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_ACT_MASK, TB_OP_READ) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK, TB_EXECUTE);
+		break;
+	case LUT_READ_METHOD_NEXT_L2UCSPA:
+		ret = rtl837x_reg_bits_write(priv, RTL8373_ITA_CTRL0_ADDR,
+					  RTL8373_ITA_CTRL0_TBL_ADDR_MASK, entry->addr);
+		if (ret)
+			return ret;
+		ret = rtl837x_reg_bits_write(priv, RTL8373_ITA_L2_CTRL_ADDR,
+					  RTL8373_ITA_L2_CTRL_PORT_NUM_MASK, entry->uc.port);
+		if (ret)
+			return ret;
+		cmd = FIELD_PREP(RTL8373_ITA_CTRL0_TBL_ADDR_MASK, entry->addr) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_TYPE_MASK, TB_TARGET_L2) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_ACT_MASK, TB_OP_READ) |
+			  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK, TB_EXECUTE);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	// execute the table cmd
+	ret = rtl837x_reg_write(priv, RTL8373_ITA_CTRL0_ADDR, cmd);
+	if (ret)
+		return ret;
+
+	// wait busy flag
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
+		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
+		  10, 10000);
+	if (ret)
+		return ret;
+
+	// check did the query hit a entry
+	ret = rtl837x_reg_bits_read(priv, RTL8373_ITA_L2_CTRL_ADDR,
+				  RTL8373_ITA_L2_CTRL_ACT_STS_MASK, &tmp);
+	if (ret)
+		return ret;
+	if (!tmp)
+		return -ENOENT; // not hit
+
+	ret = rtl837x_reg_bits_read(priv, RTL8373_ITA_L2_CTRL_ADDR,
+				  RTL8373_ITA_L2_CTRL_TBL_ADDR_MASK, &tmp);
+	if (ret)
+		return ret;
+	entry->addr = tmp;
+
+	// read the table data
+	for (int i=0; i<RTL837X_LUT_TABLE_SIZE; i++)
+	{
+		ret = rtl837x_reg_read(priv, RTL8373_ITA_READ_DATA0_ADDR(i), &tb_data[i]);
+		if (ret)
+			return ret;
+	}
+
+	// Get entry type
+	if (FIELD_GET(RTL837X_L3_D1_L3LOOKUP_MSK, tb_data[1]))
+		entry->type = LUT_TYPE_L3;
+	else if (FIELD_GET(RTL837X_L2_MC_D1_MAC0_MSK, tb_data[1]) & 1) // Check mac0 bit0
+		entry->type = LUT_TYPE_L2_MC;
+	else
+		entry->type = LUT_TYPE_L2_UC;
+
+	switch (entry->type)
+	{
+	case LUT_TYPE_L2_UC:
+		rtl837x_l2_data_to_uc(tb_data, &(entry->uc));
+		break;
+	case LUT_TYPE_L2_MC:
+		rtl837x_l2_data_to_mc(tb_data, &(entry->mc));
+		break;
+	case LUT_TYPE_L3:
+		rtl837x_data_to_l3(tb_data, &(entry->l3));
+		break;
+	}
+
+	return 0;
+}
+
+/*
+	something interesting:
+	if  multicast macaddress(mac[0] bit0 == 1) is set in the unicast entry
+	the hardware won't record this entry in the lut table
+	it will throw an error (-ENOENT)
+*/
+int rtl837x_lut_set(struct rtl837x_priv *priv, 
+                          struct rtl837x_lut_entry *entry)
+{
+	int ret;
+	u32 tmp, cmd;
+	u32 tb_data[RTL837X_LUT_TABLE_SIZE] = {0};
+
+	// check busy
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
+		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
+		  10, 10000);
+	if (ret)
+		return ret;
+
+	switch (entry->type)
+	{
+	case LUT_TYPE_L2_UC:
+		rtl837x_l2_uc_to_data(&entry->uc, tb_data);
+		break;
+	case LUT_TYPE_L2_MC:
+		rtl837x_l2_mc_to_data(&entry->mc, tb_data);
+		break;
+	case LUT_TYPE_L3:
+		rtl837x_l3_to_data(&entry->l3, tb_data);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	for (int i=0; i<RTL837X_LUT_TABLE_SIZE; i++)
+	{
+		ret = rtl837x_reg_write(priv, RTL8373_ITA_WRITE_DATA0_ADDR(i), tb_data[i]);
+		if (ret)
+			return ret;
+	}
+
+	cmd = FIELD_PREP(RTL8373_ITA_CTRL0_TLB_TYPE_MASK, TB_TARGET_L2) |
+		  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_ACT_MASK, TB_OP_WRITE) |
+		  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK, TB_EXECUTE);
+
+	// execute the table cmd
+	ret = rtl837x_reg_write(priv, RTL8373_ITA_CTRL0_ADDR, cmd);
+	if (ret)
+		return ret;
+
+	// wait busy flag
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
+		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
+		  10, 10000);
+	if (ret)
+		return ret;
+
+	// check did the query hit a entry
+	ret = rtl837x_reg_bits_read(priv, RTL8373_ITA_L2_CTRL_ADDR,
+				  RTL8373_ITA_L2_CTRL_ACT_STS_MASK, &tmp);
+	if (ret)
+		return ret;
+	if (!tmp)
+		return -ENOENT; // not hit
+
+	ret = rtl837x_reg_bits_read(priv, RTL8373_ITA_L2_CTRL_ADDR,
+				  RTL8373_ITA_L2_CTRL_TBL_ADDR_MASK, &tmp);
+	if (ret)
+		return ret;
+	entry->addr = tmp;
+
+	return 0;
+}
+
+int rtl837x_lut_del(struct rtl837x_priv *priv, 
+                          u32 addr)
+{
+	int ret;
+	u32 tmp, cmd;
+
+	// check busy
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
+		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
+		  10, 10000);
+	if (ret)
+		return ret;
+
+	ret = rtl837x_reg_bits_write(priv, RTL8373_ITA_L2_CTRL_ADDR,
+						  RTL8373_ITA_L2_CTRL_ENTRY_CLR_MASK, 1);
+	if (ret)
+		return ret;
+
+	cmd = FIELD_PREP(RTL8373_ITA_CTRL0_TBL_ADDR_MASK, addr) |
+		  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_TYPE_MASK, TB_TARGET_L2) |
+		  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_ACT_MASK, TB_OP_WRITE) |
+		  FIELD_PREP(RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK, TB_EXECUTE);
+
+	// execute the table cmd
+	ret = rtl837x_reg_write(priv, RTL8373_ITA_CTRL0_ADDR, cmd);
+	if (ret)
+		return ret;
+
+	// wait busy flag
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_ITA_CTRL0_ADDR, tmp,
+		  ((tmp & RTL8373_ITA_CTRL0_TLB_EXECUTE_MASK) == 0),
+		  10, 10000);
+	if (ret)
+		return ret;
+
+	ret = rtl837x_reg_bits_write(priv, RTL8373_ITA_L2_CTRL_ADDR,
+						  RTL8373_ITA_L2_CTRL_ENTRY_CLR_MASK, 0);
+
+	return ret;
+}
--- a/src/rtl837x.h
+++ b/src/rtl837x.h
@@ -140,6 +140,68 @@ struct rtl837x_vlan_data {
 	};
 };
 
+enum rtl837x_l2_method {
+	LUT_READ_METHOD_MAC = 0,
+	LUT_READ_METHOD_ADDRESS,
+	LUT_READ_METHOD_NEXT_ADDRESS,
+	LUT_READ_METHOD_NEXT_L2UC,
+	LUT_READ_METHOD_NEXT_L2MC,
+	LUT_READ_METHOD_NEXT_L3MC,
+	LUT_READ_METHOD_NEXT_L2L3MC,
+	LUT_READ_METHOD_NEXT_L2UCSPA,
+};
+
+enum rtl837x_lut_type {
+	LUT_TYPE_L2_UC = 1,
+	LUT_TYPE_L2_MC,
+	LUT_TYPE_L3,
+};
+
+struct rtl837x_l2_key {
+	u8 mac_addr[ETH_ALEN];
+	u16 vid_fid;
+	bool ivl;
+};
+
+struct rtl837x_l2_uc {
+	struct rtl837x_l2_key key;
+	u8 port;
+	u8 age;
+
+	bool auth; // 802.1X: not used in this driver
+	bool is_static;
+};
+
+struct rtl837x_l2_mc {
+	struct rtl837x_l2_key key;
+	u16 mbr;
+
+	bool igmp_asic;
+	u8 igmp_idx;
+};
+
+struct rtl837x_l3
+{
+	u32 sip;
+	u32 dip;
+
+	bool l3lookup;
+	u16 mbr;
+	bool igmp_asic;
+	u8 igmp_idx;
+};
+
+struct rtl837x_lut_entry {
+	enum rtl837x_lut_type type;
+	u16 addr;
+	union
+	{
+		struct rtl837x_l2_uc uc;
+		struct rtl837x_l2_mc mc;
+		struct rtl837x_l3 l3;
+	};
+};
+
 struct rtl837x_variant {
 	const struct dsa_switch_ops *ds_ops_mdio;
 	const struct rtl837x_ops *ops;
@@ -216,6 +278,14 @@ extern int rtl837x_rtl8224_sds_reg_bits_
 extern int rtl837x_vlan_set(struct rtl837x_priv *priv, struct rtl837x_vlan_data *vlan);
 extern int rtl837x_vlan_get(struct rtl837x_priv *priv, struct rtl837x_vlan_data *vlan);
 
+extern int rtl837x_lut_query(struct rtl837x_priv *priv, 
+                          enum rtl837x_l2_method method,
+                          struct rtl837x_lut_entry *entry);
+extern int rtl837x_lut_set(struct rtl837x_priv *priv, 
+                          struct rtl837x_lut_entry *entry);
+extern int rtl837x_lut_del(struct rtl837x_priv *priv, 
+                        	  u32 addr);
+
 extern int rtl837x_sds_reset_X(struct rtl837x_priv *priv, u8 sds_idx);
 extern int rtl837x_sds_reset_R(struct rtl837x_priv *priv, u8 sds_idx);
 extern int rtl837x_rtl8224_sds_reset_R(struct rtl837x_priv *priv, u8 sds_idx);
--- a/src/rtl8372n.c
+++ b/src/rtl8372n.c
@@ -9,6 +9,7 @@
 #define RTL8372N_VLAN_MEMBER_MASK 0x3FF
 #define RTL8372N_VLAN_FID_MASK 0xF
 #define RTL8372N_VLAN_MAX 4095
+#define RTL8372N_LUT_MAX 4160
 
 // TODO: this should check port is serdes mode or port mode
 #define IS_SERDES_PORT(port) (((port)==3)||((port)==8))
@@ -368,6 +369,7 @@ static int rtl8372n_vlan_update(struct rtl837x_priv *priv, int vid, u32 member,
 		return ret;
 
 	vlan4k.member |= member;
+	vlan4k.untag &= ~member;
 	vlan4k.untag |= untag;
 	vlan4k.fid = fid;
 	vlan4k.vid = vid;
@@ -376,6 +378,28 @@ static int rtl8372n_vlan_update(struct rtl837x_priv *priv, int vid, u32 member,
 	return ret;
 }
 
+/*
+ * set the port vlan egress mode
+ * If a port disable the tag rewrite
+ * frame will ignore the vlan tag config, frames can be sent out as they enter
+ * else
+ * the tag will remove/keep by vlan config
+*/
+static int rtl8372n_port_vlan_tag_rewrite(struct rtl837x_priv *priv, int port,
+		     bool enable)
+{
+	/*
+	 *	VLAN_EGRESS_TAG_MODE_ORIGINAL = 0,
+	 *	VLAN_EGRESS_TAG_MODE_KEEP_FORMAT,
+	 *	VLAN_EGRESS_TAG_MODE_PRI,
+	 *	VLAN_EGRESS_TAG_MODE_REAL_KEEP,
+	 *	VLAN_EGRESS_TAG_MODE_END
+	*/
+	return rtl837x_reg_bits_write(priv, RTL8373_VLAN_PORT_EGR_TAG_ADDR(port),
+		 RTL8373_VLAN_PORT_EGR_TAG_MODE_MASK(port), enable == true ? 0 : 1
+		);
+}
+
 static int rtl8372n_set_pvid(struct rtl837x_priv *priv, int port,
 		     u32 vid)
 {
@@ -443,11 +467,6 @@ static int rtl8372n_get_mib_counter(struct rtl837x_priv *priv,
 	}
 }
 
-static int rtl8372n_enable_vlan(struct rtl837x_priv *priv, bool enable)
-{
-    return 0;
-}
-
 static enum dsa_tag_protocol rtl8372n_get_tag_protocol(struct dsa_switch *ds,
                                                         int port,
                                                         enum dsa_tag_protocol mp)
@@ -709,7 +728,7 @@ static void rtl8372n_phylink_mac_link_down(struct phylink_config *config, unsign
 	}
 
 	if (ret) {
-		dev_err(priv->dev, "failed to disable the port(%d)\n", port);
+		dev_err(priv->dev, "MAC link down failed port(%d)\n", port);
 		return;
 	}
 }
@@ -752,7 +771,11 @@ static const struct phylink_mac_ops rtl8372n_phylink_mac_ops = {
 	.mac_link_up	= rtl8372n_phylink_mac_link_up,
 };
 
-
+/*
+ * I still don't understand how the chip GPIO register and LED configuration
+ * register work, so we use additional initialization values to configure 
+ * the chip LED and GPIO matrix or other things
+*/
 static int of_extra_init(struct dsa_switch *ds)
 {
     struct rtl837x_priv *priv = ds->priv;
@@ -810,13 +833,6 @@ static int rtl8372n_set_tag_rtl(struct dsa_switch *ds)
 		return -ENODEV;
 	}
 
-	// Set external CPU port
-	ret = rtl837x_reg_bits_write(priv, RTL8373_EXT_CPU_CTRL_ADDR,
-			  RTL8373_EXT_CPU_CTRL_PORT_MASK, cpu_dp->index
-			);
-	if (ret)
-		return ret;
-
 	// Set external CPU DSA tag insert mode
 	/*
      *	CPU_INSERT_TO_ALL = 0,
@@ -905,6 +921,10 @@ static int rtl8372n_teardown_tag_rtl(struct dsa_switch *ds)
     master_dev->wanted_features |= chip_data->csum_feature_backup;
     netdev_update_features(master_dev);
 
+	rtl837x_reg_bits_write(priv, RTL8373_L2_TBL_FLUSH_ALL_ADDR, 
+			  RTL8373_L2_TBL_FLUSH_ALL_FLUSH_ALL_MASK, 1
+			);
+
 	return 0;
 }
 
@@ -921,6 +941,13 @@ static int rtl8372n_set_tag_8021q(struct dsa_switch *ds)
 		cpu_port_mask |= BIT(dp->index);
 	}
 
+	//  bit0: internal cpu; bit1: external cpu
+	ret = rtl837x_reg_bits_write(priv, RTL8373_SVLAN_TRAP_CTRL_ADDR,
+			  RTL8373_SVLAN_TRAP_CTRL_CPU_PMSK_MASK, BIT(1)
+			);
+	if (ret)
+		return ret;
+
 	// Set S-VLAN upstream priority reference setting.
 	/*
      *	REF_INTERNAL_PRI = 0,
@@ -1024,6 +1051,10 @@ static int rtl8372n_teardown_tag_8021q(struct dsa_switch *ds)
 				  RTL8373_VS_PORT_DFLT_SVID_PORT_DFLT_SVID_MASK(dp->index), 0
 				);
 	}
+
+	rtl837x_reg_bits_write(priv, RTL8373_L2_TBL_FLUSH_ALL_ADDR, 
+			  RTL8373_L2_TBL_FLUSH_ALL_FLUSH_ALL_MASK, 1
+			);
 	return 0;
 }
 
@@ -1170,6 +1201,47 @@ static int rtl8372n_port_remove_isolation(struct rtl837x_priv *priv, int port,
 				  mask, 0);
 }
 
+static int rtl8372n_port_add_vlan_transparent(struct rtl837x_priv *priv, int port, u32 mask)
+{
+	dev_dbg(priv->dev, "[%s]: port:%d mask:0x%08x\n", __func__,
+				port, mask);
+	return rtl837x_reg_bits_write(priv, RTL8373_VLAN_PORT_EGR_TRANS_ADDR(port), 
+				  RTL8373_VLAN_PORT_EGR_TRANS_PMSK_MASK(port) & (mask << RTL8373_VLAN_PORT_EGR_TRANS_PMSK_OFFSET(port)), 0xffffffff);
+}
+
+static int rtl8372n_port_remove_vlan_transparent(struct rtl837x_priv *priv, int port, u32 mask)
+{
+	dev_dbg(priv->dev, "[%s]: port:%d mask:0x%08x\n", __func__,
+			port, mask);
+	return rtl837x_reg_bits_write(priv, RTL8373_VLAN_PORT_EGR_TRANS_ADDR(port), 
+				  RTL8373_VLAN_PORT_EGR_TRANS_PMSK_MASK(port) & (mask << RTL8373_VLAN_PORT_EGR_TRANS_PMSK_OFFSET(port)), 0);
+}
+
+static int rtl8372n_port_add_cpu_vlan_transparent(struct rtl837x_priv *priv, int port)
+{
+	struct dsa_port *cpu_dp = NULL;
+	u32 cpu_portmask = 0;
+
+	dsa_switch_for_each_cpu_port(cpu_dp, priv->ds) {
+		cpu_portmask |= BIT(cpu_dp->index);
+		rtl8372n_port_add_vlan_transparent(priv, cpu_dp->index, BIT(port));
+	}
+
+	return rtl8372n_port_add_vlan_transparent(priv, port, cpu_portmask);
+}
+
+static int rtl8372n_port_remove_cpu_vlan_transparent(struct rtl837x_priv *priv, int port)
+{
+	struct dsa_port *cpu_dp = NULL;
+	u32 cpu_portmask = 0;
+
+	dsa_switch_for_each_cpu_port(cpu_dp, priv->ds) {
+		rtl8372n_port_remove_vlan_transparent(priv, cpu_dp->index, BIT(port));
+		cpu_portmask |= BIT(cpu_dp->index);
+	}
+	return rtl8372n_port_remove_vlan_transparent(priv, port, cpu_portmask);
+}
+
 static void rtl8372n_get_strings(struct dsa_switch *ds, int port, u32 stringset,
 			 uint8_t *data)
 {
@@ -1319,6 +1391,7 @@ static void rtl8372n_get_ctrl_stats(struct dsa_switch *ds, int port,
 	mutex_unlock(&priv->mib_lock);
 }
 
+// TODO: Fail rollback
 static int rtl8372n_vlan_filtering(struct dsa_switch *ds, int port,
                                         bool vlan_filtering, struct netlink_ext_ack *extack)
 {
@@ -1336,13 +1409,14 @@ static int rtl8372n_vlan_filtering(struct dsa_switch *ds, int port,
 			);
 
 	if (vlan_filtering)
-		rtl8372n_drop_untagged(priv, port, !chip_data->pvid_enabled[port]);
+		ret = rtl8372n_drop_untagged(priv, port, !chip_data->pvid_enabled[port]);
 	else
-		rtl8372n_drop_untagged(priv, port, false);
+		ret = rtl8372n_drop_untagged(priv, port, false);
 
-    return 0;
+    return ret;
 }
 
+// TODO: Fail rollback
 static int rtl8372n_vlan_add(struct dsa_switch *ds, int port,
                             const struct switchdev_obj_port_vlan *vlan,
                             struct netlink_ext_ack *extack)
@@ -1351,7 +1425,6 @@ static int rtl8372n_vlan_add(struct dsa_switch *ds, int port,
 	bool untagged = !!(vlan->flags & BRIDGE_VLAN_INFO_UNTAGGED);
 	bool pvid = !!(vlan->flags & BRIDGE_VLAN_INFO_PVID);
     struct rtl837x_priv *priv = ds->priv;
-	struct rtl8372n *chip_data = priv->chip_data;
 
     u16 vid = vlan->vid;
 	u32 member = 0;
@@ -1363,9 +1436,9 @@ static int rtl8372n_vlan_add(struct dsa_switch *ds, int port,
         return -EINVAL;
     }
 
-	if (chip_data->dsa_tag_8021q_vid[vid])
+	if (vid_is_dsa_8021q(vid) && priv->tag_proto == DSA_TAG_PROTO_MXL862_8021Q)
     {
-        NL_SET_ERR_MSG_MOD(extack, "VLAN ID is used by DSA tag");
+        NL_SET_ERR_MSG_MOD(extack, "Range 3072-4095 reserved for dsa_8021q operation");
         return -EINVAL;
     }
 
@@ -1398,6 +1471,7 @@ static int rtl8372n_vlan_add(struct dsa_switch *ds, int port,
     return 0;
 }
 
+// TODO: Fail rollback
 static int rtl8372n_vlan_del(struct dsa_switch *ds, int port,
                                  const struct switchdev_obj_port_vlan *vlan)
 {
@@ -1432,6 +1506,7 @@ static int rtl8372n_vlan_del(struct dsa_switch *ds, int port,
 	return 0;
 }
 
+// TODO: Fail rollback
 static int
 rtl8372n_port_bridge_join(struct dsa_switch *ds, int port,
 			   struct dsa_bridge bridge,
@@ -1440,17 +1515,19 @@ rtl8372n_port_bridge_join(struct dsa_switch *ds, int port,
 {
     struct rtl837x_priv *priv = ds->priv;
 	struct dsa_port *dp;
-	unsigned int port_bitmap = 0;
+	u32 port_bitmap = 0;
 	int ret;
 
 	/* Loop over all other ports than the current one */
 	dsa_switch_for_each_user_port(dp, ds) {
-		/* Current port handled last */
-		if (dp->index == port)
-			continue;
 		/* Not on this bridge */
 		if (!dsa_port_offloads_bridge(dp, &bridge))
 			continue;
+
+		/* Current port handled last */
+		if (dp->index == port)
+			continue;
+
 		/* Join this port to each other port on the bridge */
 		ret = rtl8372n_port_add_isolation(priv, dp->index, BIT(port));
 		if (ret)
@@ -1461,30 +1538,47 @@ rtl8372n_port_bridge_join(struct dsa_switch *ds, int port,
 	dev_dbg(priv->dev, "[%s]: port(%d) isolate(0x%04x)\n", __func__,
 						  port, port_bitmap);
 
+	/*
+	 * Filter and forward the frame by vlan table
+	*/
+	ret = rtl8372n_port_remove_cpu_vlan_transparent(priv, port);
+	if (ret)
+		dev_err(priv->dev, "failed to remove port(%d)<->cpu vlan transparent err: %d\n", dp->index, ret);
+
+	if (!dsa_is_cpu_port(ds, port))
+	{
+		ret = rtl8372n_port_vlan_tag_rewrite(priv, port, true);
+		if (ret)
+			return ret;
+	}
+
 	/* Set the bits for the ports we can access */
 	ret = rtl8372n_port_add_isolation(priv, port, port_bitmap);
 	return ret;
 }
 
+// TODO: Fail rollback?
 static void
 rtl8372n_port_bridge_leave(struct dsa_switch *ds, int port,
 			    struct dsa_bridge bridge)
 {
     struct rtl837x_priv *priv = ds->priv;
 	struct dsa_port *dp;
-	unsigned int port_bitmap = 0;
+	u32 port_bitmap = 0;
 	int ret;
 	dev_dbg(priv->dev, "[%s]: %d\n", __func__,
 							  port);
 
 	/* Loop over all other ports than this one */
 	dsa_switch_for_each_user_port(dp, ds) {
-		/* Current port handled last */
-		if (dp->index == port)
-			continue;
 		/* Not on this bridge */
 		if (!dsa_port_offloads_bridge(dp, &bridge))
 			continue;
+
+		/* Current port handled last */
+		if (dp->index == port)
+			continue;
+
 		/* Remove this port from any other port on the bridge */
 		ret = rtl8372n_port_remove_isolation(priv, dp->index,
 					  BIT(port));
@@ -1494,6 +1588,21 @@ rtl8372n_port_bridge_leave(struct dsa_switch *ds, int port,
 		port_bitmap |= BIT(dp->index);
 	}
 
+	/*
+	 * When the port is not in the bridge, in order 
+	 * to allow all VLAN tags to be accepted, 
+	 * VLAN transparent transmission is set
+	*/
+	ret = rtl8372n_port_add_cpu_vlan_transparent(priv, port);
+	if (ret)
+		dev_err(priv->dev, "failed to add port(%d)<->cpu vlan transparent err: %d\n", dp->index, ret);
+
+	/*
+	 * Set the hardware do not add/remove/edit the vlan tag
+	 * The VLAN remains completely unchanged when the frame enters and exits
+	*/
+	rtl8372n_port_vlan_tag_rewrite(priv, port, false);
+
 	/* Clear the bits for the ports we can not access, leave ourselves */
 	rtl8372n_port_remove_isolation(priv, port, port_bitmap);
 }
@@ -1524,6 +1633,309 @@ static void rtl8372n_port_disable(struct dsa_switch *ds, int port)
 	priv->ops->phy_write_c45(priv, port, 31, 0xa610, 0x2858);
 }
 
+static int
+rtl8372n_port_fdb_static_add(struct rtl837x_priv *priv, int port,
+		    const unsigned char *addr, u16 vid)
+{
+	int ret;
+	struct rtl837x_lut_entry entry = {0};
+
+	memcpy(entry.uc.key.mac_addr, addr, ETH_ALEN);
+	entry.type = LUT_TYPE_L2_UC;
+	entry.uc.key.vid_fid = vid;
+	entry.uc.key.ivl = true;
+
+	// age shoule greater than 0, so that the static entry will not be flushed by hardware
+	entry.uc.age = 6;
+	entry.uc.port = port;
+	entry.uc.is_static = true;
+
+	ret = rtl837x_lut_set(priv, &entry);
+	if (ret == -ENOENT)
+	{
+		dev_dbg(priv->dev, "[%s]:addfailed mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d\n", __func__,
+						addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+						port, vid);
+		return -ENOSPC;
+	}
+	dev_dbg(priv->dev, "[%s]:addsucceed mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d addr: %04d\n", __func__,
+					addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+					port, vid, entry.addr);
+	return ret;
+}
+
+static int
+rtl8372n_port_fdb_static_del(struct rtl837x_priv *priv,
+		    const unsigned char *addr, u16 vid)
+{
+	int ret;
+	struct rtl837x_lut_entry entry = {0};
+
+	memcpy(entry.uc.key.mac_addr, addr, ETH_ALEN);
+	entry.type = LUT_TYPE_L2_UC;
+	entry.uc.key.vid_fid = vid;
+	entry.uc.key.ivl = true;
+
+	// query the entry in lut table
+	ret = rtl837x_lut_query(priv, LUT_READ_METHOD_MAC, &entry);
+	if (ret == -ENOENT)
+	{
+		dev_dbg(priv->dev, "[%s]:notfound mac:%02X:%02X:%02X:%02X:%02X:%02X vid:%04d\n", __func__,
+						addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+						vid);
+		// Not found in lut table
+		// just return success
+		return 0;
+	}
+	dev_dbg(priv->dev, "[%s]:deleted mac:%02X:%02X:%02X:%02X:%02X:%02X vid:%04d addr: %04d\n", __func__,
+					addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+					vid, entry.addr);
+	return rtl837x_lut_del(priv, entry.addr);
+}
+
+static int
+rtl8372n_port_fdb_add(struct dsa_switch *ds, int port,
+		    const unsigned char *addr, u16 vid,
+		    struct dsa_db db)
+{
+	struct rtl837x_priv *priv = ds->priv;
+	int ret;
+
+	if (db.type != DSA_DB_PORT && db.type != DSA_DB_BRIDGE)
+		return -EOPNOTSUPP;
+
+	dev_dbg(priv->dev, "[%s]:type: %s port:%d vid:%04d\n", __func__,
+					db.type == DSA_DB_PORT ? "DSA_DB_PORT" : "DSA_DB_BRIDGE",
+					port, vid);
+
+	if (db.type == DSA_DB_BRIDGE && priv->tag_proto == DSA_TAG_PROTO_MXL862_8021Q)
+	{
+		//Because the CPU port can only learn 802.1q tags, So VLAN tags are ignored
+		if (vid != 0)
+			return 0;
+		/*
+		* When the DSA tag protocol is DSA_TAG_PROTO_MXL862_8021Q, the L2 VLAN
+		* learned on the CPU port is the 802.1q tag rather than the port's own
+		* VLAN ID. Therefore, write the 802.1q tags of all ports into the
+		* static FDB entries.
+		*/
+		struct dsa_port *dp;
+		dsa_switch_for_each_user_port(dp, ds) {
+			/* Add static fdb entry */
+			ret = rtl8372n_port_fdb_static_add(priv, port, addr, dsa_tag_8021q_standalone_vid(dp));
+			if (ret)
+				return ret;
+		}
+	} else
+	{
+		return rtl8372n_port_fdb_static_add(priv, port, addr, vid);
+	}
+	return 0;
+}
+
+static int
+rtl8372n_port_fdb_del(struct dsa_switch *ds, int port,
+		    const unsigned char *addr, u16 vid,
+		    struct dsa_db db)
+{
+	struct rtl837x_priv *priv = ds->priv;
+	int ret;
+
+	if (db.type != DSA_DB_PORT && db.type != DSA_DB_BRIDGE)
+		return -EOPNOTSUPP;
+
+	if (dsa_fdb_present_in_other_db(ds, port, addr, vid, db))
+		return 0;
+
+	if (db.type == DSA_DB_BRIDGE && priv->tag_proto == DSA_TAG_PROTO_MXL862_8021Q)
+	{
+		if (vid != 0)
+			return 0;
+		struct dsa_port *dp;
+		dsa_switch_for_each_user_port(dp, ds) {
+			/* Del static fdb entry */
+			ret = rtl8372n_port_fdb_static_del(priv, addr, dsa_tag_8021q_standalone_vid(dp));
+			if (ret)
+				return ret;
+		}
+	} else
+	{
+		return rtl8372n_port_fdb_static_del(priv, addr, vid);
+	}
+	return 0;
+}
+
+static int
+rtl8372n_port_fdb_dump(struct dsa_switch *ds, int port,
+		     dsa_fdb_dump_cb_t *cb, void *data)
+{
+	struct rtl837x_priv *priv = ds->priv;
+	int ret;
+	struct rtl837x_lut_entry entry = {0};
+
+	for (int i = 0; i < RTL8372N_LUT_MAX; i++) {
+		entry.addr = i;
+		entry.uc.port = port;
+		ret = rtl837x_lut_query(priv, LUT_READ_METHOD_NEXT_L2UCSPA, &entry);
+		if (ret)
+			break;
+		if (entry.addr < i)
+			break;
+		/* We need to hide the dsa_8021q VLANs from the user. */
+		if (vid_is_dsa_8021q(entry.uc.key.vid_fid) && priv->tag_proto == DSA_TAG_PROTO_MXL862_8021Q)
+			entry.uc.key.vid_fid = 0;
+		ret = cb(entry.uc.key.mac_addr, entry.uc.key.vid_fid, entry.uc.is_static,
+				data);
+		if (ret < 0)
+			break;
+		i = entry.addr;
+	}
+	return 0;
+}
+
+static int
+rtl8372n_port_mdb_add(struct dsa_switch *ds, int port,
+		    const struct switchdev_obj_port_mdb *mdb,
+		    struct dsa_db db)
+{
+	struct rtl837x_priv *priv = ds->priv;
+	int ret;
+	const u8 *addr = mdb->addr;
+	u16 vid = mdb->vid;
+	struct rtl837x_lut_entry entry = {0};
+
+	memcpy(entry.mc.key.mac_addr, addr, ETH_ALEN);
+	entry.type = LUT_TYPE_L2_MC;
+	entry.mc.key.vid_fid = vid;
+	entry.mc.key.ivl = true;
+	ret = rtl837x_lut_query(priv, LUT_READ_METHOD_MAC, &entry);
+	if (ret != -ENOENT && ret != 0)
+		return ret;
+
+	if (ret == 0)
+	{
+		dev_dbg(priv->dev, "[%s]:found mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d addr:%04d\n", __func__,
+						addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+						port, vid, entry.addr);
+		if (entry.type != LUT_TYPE_L2_MC)
+		{
+			dev_err(priv->dev, "[%s]: got an unexpect entry from lut table, this should not happen\n", __func__);
+			return -EINVAL;
+		}
+	}
+
+	entry.mc.mbr |= BIT(port);
+	ret = rtl837x_lut_set(priv, &entry);
+	if (ret == -ENOENT)
+	{
+		dev_dbg(priv->dev, "[%s]:addfailed mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d\n", __func__,
+						addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+						port, vid);
+		return -ENOSPC;
+	}
+	dev_dbg(priv->dev, "[%s]:addsucceed mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d addr: %04d\n", __func__,
+					addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+					port, vid, entry.addr);
+	return ret;
+}
+
+static int
+rtl8372n_port_mdb_del(struct dsa_switch *ds, int port,
+		    const struct switchdev_obj_port_mdb *mdb,
+		    struct dsa_db db)
+{
+	struct rtl837x_priv *priv = ds->priv;
+	int ret;
+	const u8 *addr = mdb->addr;
+	u16 vid = mdb->vid;
+	struct rtl837x_lut_entry entry = {0};
+	dev_dbg(priv->dev, "[%s]:mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d\n", __func__,
+					addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+					port, vid);
+
+	memcpy(entry.mc.key.mac_addr, addr, ETH_ALEN);
+	entry.type = LUT_TYPE_L2_MC;
+	entry.mc.key.vid_fid = vid;
+	entry.mc.key.ivl = true;
+	ret = rtl837x_lut_query(priv, LUT_READ_METHOD_MAC, &entry);
+	if (ret == -ENOENT)
+		return 0;
+
+	if (ret == 0)
+	{
+		dev_dbg(priv->dev, "[%s]:found mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d addr:%04d\n", __func__,
+						addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+						port, vid, entry.addr);
+		if (entry.type != LUT_TYPE_L2_MC)
+		{
+			dev_err(priv->dev, "[%s]: got an unexpect entry from lut table, this should not happen\n", __func__);
+			return -EINVAL;
+		}
+	}else
+		return ret;
+
+	entry.mc.mbr &= ~BIT(port);
+	if (!entry.mc.mbr)
+	{
+		return rtl837x_lut_del(priv, entry.addr);
+	}
+
+	ret = rtl837x_lut_set(priv, &entry);
+	if (ret == -ENOENT)
+	{
+		dev_dbg(priv->dev, "[%s]:delfailed mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d\n", __func__,
+						addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+						port, vid);
+		return -ENOSPC;
+	}
+	dev_dbg(priv->dev, "[%s]:delsuceed mac:%02X:%02X:%02X:%02X:%02X:%02X port:%d vid:%04d addr: %04d\n", __func__,
+					addr[0], addr[1], addr[2], addr[3], addr[4], addr[5],
+					port, vid, entry.addr);
+	return ret;
+}
+
+enum RTL8373_FLUSHMODE
+{
+    FLUSHMDOE_PORT = 0,
+    FLUSHMDOE_VID,
+    FLUSHMDOE_FID,
+    FLUSHMDOE_END,
+};
+
+static void rtl8372n_port_fast_age(struct dsa_switch *ds, int port)
+{
+	struct rtl837x_priv *priv = ds->priv;
+	int ret;
+	u32 tmp;
+
+	// check busy
+	ret = regmap_read_poll_timeout(priv->map, RTL8373_L2_TBL_FLUSH_CMD_ADDR, tmp,
+		  ((tmp & RTL8373_L2_TBL_FLUSH_CMD_FLUSH_BUSY_MASK) == 0),
+		  10, 1000);
+	if (ret)
+		return;
+
+	ret = rtl837x_reg_bits_write(priv, RTL8373_L2_TBL_FLUSH_MODE_ADDR,
+					  RTL8373_L2_TBL_FLUSH_MODE_FLUSH_MODE_MASK, FLUSHMDOE_PORT);
+	if (ret)
+		return;
+	ret = rtl837x_reg_bits_write(priv, RTL8373_L2_TBL_FLUSH_CMD_ADDR,
+					  RTL8373_L2_TBL_FLUSH_CMD_FLUSH_PMSK_MASK, BIT(port));
+	if (ret)
+		return;
+
+	ret = rtl837x_reg_bits_write(priv, RTL8373_L2_TBL_FLUSH_CMD_ADDR,
+					  RTL8373_L2_TBL_FLUSH_CMD_FLUSH_ACT_MASK, 1);
+	if (ret)
+		return;
+
+	// check busy
+	regmap_read_poll_timeout(priv->map, RTL8373_L2_TBL_FLUSH_CMD_ADDR, tmp,
+		  ((tmp & RTL8373_L2_TBL_FLUSH_CMD_FLUSH_BUSY_MASK) == 0),
+		  10, 1000);
+	return;
+}
+
 enum RTL8373_MSTP_STATE
 {
     MSTP_DISABLE = 0,
@@ -1581,14 +1993,14 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 	struct rtl8372n *chip_data = priv->chip_data;
 	struct dsa_port *cpu_dp = NULL;
 	struct dsa_port *dp;
-	u32 downports_mask = 0;
+	u32 downports_mask = 0,
+		cpu_port_mask = 0;
 
 	int cpu_dp_cnt = 0;
-	dsa_switch_for_each_port(dp, ds) {
-		if (dsa_port_is_cpu(dp)) {
-			cpu_dp = dp;
-			cpu_dp_cnt++;
-		}
+	dsa_switch_for_each_cpu_port(dp, ds) {
+		cpu_port_mask |= BIT(dp->index);
+		cpu_dp = dp;
+		cpu_dp_cnt++;
 	}
 
 	// TODO: muilt CPU port support
@@ -1617,8 +2029,6 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 	chip_data->pcs[8].priv = priv;
 	chip_data->pcs[8].index = 8;
 
-    dev_info(priv->dev,"Start init RTL8372N Switch\n");
-
 	// set port 3 and port 8 as serdes port
 	rtl837x_reg_bits_write(priv, RTL8373_SMI_MAC_TYPE_CTRL_ADDR, 
 			 RTL8373_SMI_MAC_TYPE_CTRL_MAC_PORT8_TYPE_MASK | RTL8373_SMI_MAC_TYPE_CTRL_MAC_PORT3_TYPE_MASK,
@@ -1638,6 +2048,9 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 			 0b111
 			);
 
+	//  puts "Power down PHY 4~7"
+	rtl837x_phys_write_c45(priv, 0xF0, 31, 0xa610, 0x2858);
+
 	if (of_property_read_bool(np, "sds0-rx-swap"))
 	{
 		rtl837x_sds_reg_bits_write(priv, 0, 0, 0, 0x200, 1); //#SDS0RX PN swap
@@ -1712,36 +2125,17 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 				 0x596A
 				); //#TX_POLARITY_SWAP
 
-	//  puts "Power down PHY 4~7"
-	rtl837x_phys_write_c45(priv, 0xF0, 31, 0xa610, 0x2858);
-
-	//## ---------------------------Patch MAC--------------------------
 	//#cfg_FWD_INVLD_MAC_CTRL_EN,cfg_FWD_UNKN_OPCODE_EN
 	rtl837x_reg_bits_write(priv, RTL8373_MAC_L2_GLOBAL_CTRL0_ADDR,
 			 RTL8373_MAC_L2_GLOBAL_CTRL0_FWD_UNKN_OPCODE_EN_MASK | RTL8373_MAC_L2_GLOBAL_CTRL0_FWD_INVLD_MAC_CTRL_EN_MASK,
 			 0b11
 			);
 
-	for(int i=3; i<9; i++)
-	{
-		rtl837x_reg_bits_write(priv, RTL8373_MAC_L2_PORT_CTRL_ADDR(i),
-		 RTL8373_MAC_L2_PORT_CTRL_RX_CHK_CRC_EN_MASK, 1
-		);
-		rtl837x_reg_bits_write(priv, RTL8373_MAC_L2_PORT_CTRL_ADDR(i),
-		 RTL8373_MAC_L2_PORT_CTRL_CLOCK_SWITCH_MASK, 1
-		);
-	}
-
 	// #RS_LINK_FAULT_INDI_OFF=1 disable link fault flag, resolve port4-port7 linkdown dsc expand issue
     rtl837x_reg_bits_write(priv, RTL8373_RS_LAYER_CONFIG_ADDR,
 		 RTL8373_RS_LAYER_CONFIG_RS_LINK_FAULT_INDI_OFF_MASK, 1
 		);
 
-	for(int i=0; i<10; i++)
-    {
-        rtl837x_reg_write(priv, RTL8373_FC_PORT_ACT_CTRL_ADDR(i), 0x1050);
-    }
-
 	rtl837x_reg_bits_write(priv, RTL8373_DW8051_CFG_ADDR,
 			 RTL8373_DW8051_CFG_DW8051_READY_MASK, 1
 			);
@@ -1754,14 +2148,19 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 	}
 #endif
 
-	//  puts "Power up PHY 4~7"
-    rtl837x_phys_write_c45(priv, 0xF0 ,31,0xa610,0x2058);
     //RTL8372/RTL8372N/RTL8366U set polling mask 0x1f8, port 3/8 from serdes need config bit8=1
     rtl837x_reg_bits_write(priv, RTL8373_SMI_GLB_CTRL_ADDR,
 		 RTL8373_SMI_GLB_CTRL_SMI_POLLING_MASK_MASK, 0x1f8
 		);
 	msleep(5);
 
+
+	dsa_switch_for_each_port(dp, ds) {
+		ret = rtl8372n_port_remove_vlan_transparent(priv, dp->index, 0xffffffff);
+		if(ret)
+			return ret;
+	}
+
 	of_extra_init(ds);
 
     ret = rtl8372n_setup_mdio(priv);
@@ -1778,6 +2177,22 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 	dsa_switch_for_each_port(dp, ds) {
 		int port = dp->index;
 
+        ret = rtl837x_reg_write(priv, RTL8373_FC_PORT_ACT_CTRL_ADDR(port), 0x1050);
+		if (ret)
+			return ret;
+
+		ret = rtl837x_reg_bits_write(priv, RTL8373_MAC_L2_PORT_CTRL_ADDR(port),
+			  RTL8373_MAC_L2_PORT_CTRL_RX_CHK_CRC_EN_MASK, 1
+			);
+		if (ret)
+			return ret;
+
+		ret = rtl837x_reg_bits_write(priv, RTL8373_MAC_L2_PORT_CTRL_ADDR(port),
+			  RTL8373_MAC_L2_PORT_CTRL_CLOCK_SWITCH_MASK, 1
+			);
+		if (ret)
+			return ret;
+
 		rtl8372n_port_stp_state_set(ds, port, BR_STATE_DISABLED);
 
 		/* Start with all port completely isolated */
@@ -1788,11 +2203,6 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 		if (dsa_port_is_unused(dp))
 			continue;
 
-    	// Disable per-port l2 learning
-		ret = rtl837x_reg_write(priv, RTL8373_L2_LRN_PORT_CONSTRT_CTRL_ADDR(port), 0);
-		if (ret)
-			return ret;
-
 		// FORWARD:0, DROP:1, TO_CPU:2
 		ret = rtl837x_reg_bits_write(priv, RTL8373_L2_LRN_PORT_CONSTRT_ACT_ADDR,
 			 RTL8373_L2_LRN_PORT_CONSTRT_ACT_LRN_ACT_MASK, 0
@@ -1800,17 +2210,7 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 		if (ret)
 			return ret;
 
-		// What fuck is this?
-		/*
-		 *	VLAN_EGRESS_TAG_MODE_ORIGINAL = 0,
-		 *	VLAN_EGRESS_TAG_MODE_KEEP_FORMAT,
-		 *	VLAN_EGRESS_TAG_MODE_PRI,
-		 *	VLAN_EGRESS_TAG_MODE_REAL_KEEP,
-		 *	VLAN_EGRESS_TAG_MODE_END
-		*/
-		ret = rtl837x_reg_bits_write(priv, RTL8373_VLAN_PORT_EGR_TAG_ADDR(port),
-			 RTL8373_VLAN_PORT_EGR_TAG_MODE_MASK(port), 1
-			);
+		ret = rtl8372n_port_vlan_tag_rewrite(priv, port, false);
 		if (ret)
 			return ret;
 
@@ -1818,6 +2218,7 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 		ret = rtl8372n_drop_untagged(priv, port, false);
 		if (ret)
 			return ret;
+
 		// Disable Ingress filter
 		ret = rtl837x_reg_bits_write(priv, RTL8373_VLAN_PORT_IGR_FLTR_ADDR(port),
 			  RTL8373_VLAN_PORT_IGR_FLTR_IGR_FLTR_ACT_MASK(port), 0
@@ -1842,6 +2243,11 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 
 		if (!dsa_port_is_user(dp))
 			continue;
+
+		ret = rtl8372n_port_add_cpu_vlan_transparent(priv, port);
+		if (ret)
+			return ret;
+
 		/* Forward only to the CPU */
 		ret = rtl8372n_port_set_isolation(priv, dp->index,
 						   BIT(cpu_dp->index));
@@ -1853,14 +2259,146 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 
 	ret = rtl8372n_port_set_isolation(priv, cpu_dp->index,
 						downports_mask);
+	if (ret)
+		return ret;
+
+	/*
+	 * Warning! 
+	 * The following understanding may be incorrect
+	 * 
+	 * !!! The annotation is outdated !!!
+	*/
 
-	// Disable l2 learning
-	ret = rtl837x_reg_bits_write(priv, RTL8373_L2_LRN_CONSTRT_CTRL_ADDR,
-			  RTL8373_L2_LRN_CONSTRT_CTRL_CONSTRT_NUM_MASK, 0
+	/*
+	 * So What is this.
+	 * We use vlan1 to forward the untag frame in the bridge
+	 * If the CPU wants to send a frame to a port and the frame does not have a VLAN tag
+	 * The switch will insert VLAN1 into the frame and remove the VLAN1 when sending the frame
+	 * The same applies to frames entering through ports
+	 * 
+	 * 
+	 * If a frame send from CPU (CPU->switch) whithout cvid only with a DSA TAG
+	 * ================================================================================
+	 * --------------------------------------------------------------------------------
+	 *                                 SVLAN Process
+	 *  |DMAC|SMAC|DSA TAG(SVLAN)|...| -------------> |DMAC|SMAC|...| ----..
+	 *              remove dsa tag (maybe? or remove the tag before the frame sent out?)
+	 *                              mark destination port
+	 *                                  
+	 * --------------------------------------------------------------------------------         
+	 *       CVLAN process                            Send to dest port       
+	 * ..---------------------> |DMAC|SMAC|CVLAN|...| -----------------> |DMAC|SMAC|...|
+	 *  frame with out CVLAN tag                     remove the CVLAN tag
+	 * mark the cpuport pvid(1)
+	 * we use vid 1(1) to forward the
+	 *     no CVLAN frame
+	 * --------------------------------------------------------------------------------
+	 * ================================================================================
+	 * 
+	 * If a frame send from CPU (CPU->switch) whith DSA tag(SVLAN) and CVLAN
+	 * ================================================================================
+	 * --------------------------------------------------------------------------------
+	 *                                       SVLAN Process
+	 *  |DMAC|SMAC|DSA TAG(SVLAN)|CVLAN|...| -------------> |DMAC|SMAC|CVLAN|...| ----..
+	 *                                       remove dsa tag
+	 *                                    mark destination port
+	 * --------------------------------------------------------------------------------         
+	 *       CVLAN process                            Send to dest port       
+	 * ..---------------------> |DMAC|SMAC|CVLAN|...| -----------------> |DMAC|SMAC|CVLAN(may not exist)|...|
+	 *  frame already with CVLAN tag             remove or keep the CVLAN tag
+	 * 
+	 * --------------------------------------------------------------------------------
+	 * ================================================================================
+	 * 
+	 * If a frame send to CPU (switch->CPU) whithout vlan tag
+	 * ================================================================================
+	 * --------------------------------------------------------------------------------
+	 *                  Port based vlan tag
+	 *  |DMAC|SMAC|...| -------------------> |DMAC|SMAC|CVLAN|...| ----...
+	 *               add the port based vlan id
+	 * 
+	 * --------------------------------------------------------------------------------         
+	 *   SVLAN process
+	 *  ..-----------------> |DMAC|SMAC|DSA TAG(SVLAN)|CVLAN|...|
+	 * mark the src port svid
+	 * 
+	 * --------------------------------------------------------------------------------
+	 * ================================================================================
+	 * 
+	 * If a frame send to CPU (switch->CPU) whit vlan tag
+	 * ================================================================================
+	 * --------------------------------------------------------------------------------
+	 *                        Port Ingerss check
+	 *  |DMAC|SMAC|CVLAN|...| -------------------> |DMAC|SMAC|CVLAN|...| ----...
+	 *                         drop or forward
+	 *                         
+	 * --------------------------------------------------------------------------------         
+	 *   SVLAN process
+	 *  ..-----------------> |DMAC|SMAC|DSA TAG(SVLAN)|CVLAN|...|
+	 * mark the src port svid
+	 * 
+	 * --------------------------------------------------------------------------------
+	 * ================================================================================
+	*/
+
+	/*
+	 * So when a port is not in the bridge
+	 * what will happens
+	 * 
+	 * when the port is not in the bridge, it will enable vlan transparent on it
+	 * This way, the port will not be restricted by the VLAN table
+	 * the frame will only add/remove the DSA TAG(SVLAN), and there will be no changes to CVLAN
+	 * 
+	 * If a frame send to CPU (switch->CPU)
+	 * ================================================================================
+	 * --------------------------------------------------------------------------------
+	 *                                Port Ingerss check
+	 *  |DMAC|SMAC|CVLAN(or not)|...| -------------------> |DMAC|SMAC|CVLAN(or not)|...| ----...
+	 *                                 check is disabled
+	 *                               all frames can came in
+	 * --------------------------------------------------------------------------------         
+	 *   SVLAN process
+	 *  ..-----------------> |DMAC|SMAC|DSA TAG(SVLAN)|CVLAN(or not)|...|
+	 * mark the src port svid
+	 * 
+	 * --------------------------------------------------------------------------------
+	 * ================================================================================
+	 * 
+	 * If a frame send from CPU (CPU->switch)
+	 * ================================================================================
+	 * --------------------------------------------------------------------------------
+	 *                                               SVLAN Process
+	 *  |DMAC|SMAC|DSA TAG(SVLAN)|CVLAN(or not)|...| -------------> |DMAC|SMAC|CVLAN(or not)|...| ----..
+	 *                                              remove dsa tag
+	 *                                           mark destination port
+	 * --------------------------------------------------------------------------------         
+	 *       CVLAN process                                    Send to dest port       
+	 * ..---------------------> |DMAC|SMAC|CVLAN(or not)|...| -----------------> |DMAC|SMAC|CVLAN((or not))|...|
+	 *        do nothing
+	 * --------------------------------------------------------------------------------
+	 * ================================================================================
+	*/
+
+	// Set external CPU port
+	ret = rtl837x_reg_bits_write(priv, RTL8373_EXT_CPU_CTRL_ADDR,
+			  RTL8373_EXT_CPU_CTRL_PORT_MASK, cpu_dp->index
 			);
 	if (ret)
 		return ret;
 
+	ret = rtl837x_reg_bits_write(priv, RTL8373_L2_TBL_FLUSH_ALL_ADDR,
+				  RTL8373_L2_TBL_FLUSH_ALL_FLUSH_ALL_MASK, 1);
+	if (ret)
+		return ret;
+
+	ret = rtl837x_reg_write(priv, RTL8373_VLAN_L2_LRN_DIS_ADDR(0), 0);
+	if (ret)
+		return ret;
+
+	ret = rtl837x_reg_write(priv, RTL8373_VLAN_L2_LRN_DIS_ADDR(1), 0);
+	if (ret)
+		return ret;
+
 	// FORWARD:0, DROP:1, TO_CPU:2
 	ret = rtl837x_reg_bits_write(priv, RTL8373_L2_LRN_PORT_CONSTRT_ACT_ADDR,
 			  RTL8373_L2_LRN_PORT_CONSTRT_ACT_LRN_ACT_MASK, 0
@@ -1876,15 +2414,9 @@ static int rtl8372n_setup(struct dsa_switch *ds)
 		return ret;
 
 	// Disable vlan leaky
-	ret = rtl837x_reg_bits_write(priv, RTL8373_MIR_CTRL_ADDR,
-			  RTL8373_MIR_CTRL_MIR_TX_VLAN_LKY_MASK | RTL8373_MIR_CTRL_MIR_RX_VLAN_LKY_OFFSET,
-			  0
-			);
-	if (ret)
-		return ret;
-
 	// Disable isolate leaky
 	ret = rtl837x_reg_bits_write(priv, RTL8373_MIR_CTRL_ADDR,
+			  RTL8373_MIR_CTRL_MIR_TX_VLAN_LKY_MASK | RTL8373_MIR_CTRL_MIR_RX_VLAN_LKY_MASK |
 			  RTL8373_MIR_CTRL_MIR_TX_ISOLATE_LKY_MASK | RTL8373_MIR_CTRL_MIR_RX_ISOLATE_LKY_MASK,
 			  0
 			);
@@ -1932,6 +2464,13 @@ static const struct dsa_switch_ops rtl8372n_switch_ops_mdio = {
 	.port_bridge_join = rtl8372n_port_bridge_join,
 	.port_bridge_leave = rtl8372n_port_bridge_leave,
 
+	.port_fdb_add  = rtl8372n_port_fdb_add,
+	.port_fdb_del  = rtl8372n_port_fdb_del,
+	.port_fdb_dump = rtl8372n_port_fdb_dump,
+	.port_mdb_add  = rtl8372n_port_mdb_add,
+	.port_mdb_del  = rtl8372n_port_mdb_del,
+    .port_fast_age = rtl8372n_port_fast_age,
+
 	.tag_8021q_vlan_add = rtl8372n_tag_8021q_vlan_add,
 	.tag_8021q_vlan_del = rtl8372n_tag_8021q_vlan_del,
 
@@ -1944,7 +2483,6 @@ static const struct rtl837x_ops rtl8372n_ops = {
 	.get_vlan_4k	= rtl8372n_get_vlan_4k,
 	.set_vlan_4k	= rtl8372n_set_vlan_4k,
 	.get_mib_counter = rtl8372n_get_mib_counter,
-	.enable_vlan	= rtl8372n_enable_vlan,
 
 	.phy_read_c22   = rtl837x_phy_read_c22,
 	.phy_write_c22  = rtl837x_phy_write_c22,
